include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -85,215 +85,284 @@ DEAL_II_NAMESPACE_OPEN
+    interpolation.reinit(vec);
+    weights.reinit(vec);
+
+    // We use an FEValues object to transform all generalized support
+    // points from the unit cell to the real cell coordinates. Thus,
+    // initialize a quadrature with all generalized support points and
+    // create an FEValues object with it.
+
+    hp::QCollection<dim> support_quadrature;
     for (unsigned int fe_index=0; fe_index<fe.size(); ++fe_index)
       {
-        for (unsigned int component_index = 0; component_index < n_components; ++component_index)
-          {
-            if (component_mask[component_index] == true)
-              {
-                Assert ((fe[fe_index].base_element(fe[fe_index].component_to_base_index
-                                                   (component_index).first).has_support_points()) ||
-                        (fe[fe_index].base_element(fe[fe_index].component_to_base_index
-                                                   (component_index).first).dofs_per_cell == 0),
-                        ExcNonInterpolatingFE());
-              }
-          }
+        const auto &points = fe[fe_index].get_generalized_support_points();
+        support_quadrature.push_back(Quadrature<dim>(points));
       }
 
-    // Find the support points on a cell that are mentioned multiple times, and
-    // only add each once.  Each multiple point gets to know the dof index of
-    // its representative point by the dof_to_rep_dof_table.
+    hp::MappingCollection<dim, spacedim> mapping_collection(mapping);
 
-    // the following vector collects all unit support points p[i],
-    // 0<=i<fe.dofs_per_cell, for which unit_support_point(i) is unique
-    // (representative). The position of a support point within this vector is
-    // called the rep index.
-    std::vector<std::vector<Point<dim> > > rep_unit_support_points (fe.size());
-    // the following table converts a dof i
-    // to the rep index.
-    std::vector<std::vector<types::global_dof_index> > dof_to_rep_index_table(fe.size());
+    hp::FEValues<dim, spacedim> fe_values(
+      mapping_collection,
+      fe,
+      support_quadrature,
+      update_quadrature_points);
+
+    // An extra FEValues object to compute jacobians.
+    // Only re-initialized in case of Hcurl or Hdiv conforming elements,
+    // i.e. if we really need the information.
+    hp::FEValues<dim, spacedim> fe_values_jacobians(
+      mapping_collection,
+      fe,
+      support_quadrature,
+      update_jacobians | update_inverse_jacobians);
 
-    std::vector<unsigned int> n_rep_points (fe.size(), 0);
+    //
+    // Now loop over all locally owned, active cells.
+    //
 
-    for (unsigned int fe_index=0; fe_index<fe.size(); ++fe_index)
+    for (auto cell : dof_handler.active_cell_iterators())
       {
-        for (unsigned int i=0; i<fe[fe_index].dofs_per_cell; ++i)
+        // If this cell is not locally owned, do nothing.
+        if (! cell->is_locally_owned())
+          continue;
```

Comment:
Not really. The other processor will add to the global vector just as we do. This follows the same principle of assembly. We don't assemble on ghost cells (because our neighbor will)...

Reply:
Ah, yes, of course. Doh.